### PR TITLE
feat(azure-cosmosdb): add session context and retrieve all sessions for a user

### PR DESCRIPTION
**Pull request by sinedied**
- Added generic `context` object saved along the session. This can be used for example to give a title or save any additional properties along with a session.
- Added method to retrieve all sessions for a user (without the messages)
- Update to latest CosmosDB SDK (had to fix a few types in vector search integrations)

@jacoblee93 Not directly related to the PR, but I've found some memory integration that uses `RunnableWithMessageHistory` like ours, and other using `BufferMemory` which one is the recommended approach? Thanks

### CODE REVIEW

**Review by aditishree1 — COMMENTED** [review]
nit: Should we add this in example or doc-comment to let the user know that they have the option to add context in input?

Otherwise, this PR LGTM.

**Comment by aditishree1** [comment]
hey @sinedied, Jacob suggested to prefer RunnableWithMessageHistory. 
 https://github.com/langchain-ai/langchainjs/pull/7057#discussion_r1837250292

**Comment by sinedied** [comment]
> nit: Should we add this in example or doc-comment to let the user know that they have the option to add context in input?
> 
> Otherwise, this PR LGTM.

Good callout! I'll update the example with the new methods.

**Comment by sinedied** [comment]
@aditishree1 I completed the code example with the new APIs usage.

I also had to fix a few types in the vector search integration due to the updated Cosmos SDK.

**Review by aditishree1 — APPROVED** [review]

**Comment by jacoblee93** [comment]
> which one is the recommended approach?

We have been leaning more and more on LangGraph as the memory abstraction of choice (you are defining a state with whatever keys you want) but if you need to stay in LangChain then `RunnableWithMessageHistory`

File: libs/langchain-azure-cosmosdb/src/azure_cosmosdb_nosql.ts [hunk 1]
```diff
@@ -195,7 +197,7 @@ export class AzureCosmosDBNoSQLVectorStore extends VectorStore {
         {
           path: "/vector",
           type: "quantizedFlat",
-        },
+        } as VectorIndex,
```
Comment (jacoblee93) [inline libs/langchain-azure-cosmosdb/src/azure_cosmosdb_nosql.ts:200]: Maybe `as const` works but not a huge deal

File: libs/langchain-azure-cosmosdb/src/chat_histories.ts [hunk 2]
```diff
@@ -201,4 +218,36 @@ export class AzureCosmsosDBNoSQLChatMessageHistory extends BaseListChatMessageHi
       await this.container.item(userSession.id, userId).delete();
     }
   }
+
+  async getAllSessionsForUser(userId: string): Promise<ChatSession[]> {
```
Comment (jacoblee93) [inline libs/langchain-azure-cosmosdb/src/chat_histories.ts:222]: Should this be a util instead? Doesn't seem to really fit on this class

File: libs/langchain-azure-cosmosdb/src/azure_cosmosdb_nosql.ts (hunk 1 above)
Comment (sinedied) [reply libs/langchain-azure-cosmosdb/src/azure_cosmosdb_nosql.ts:200]: Tried it but it still complained 😞

**Comment by sinedied** [comment]
@jacoblee93 @aditishree1 I've reworked how the helper methods works so it makes more sense, and fixed handling of existing context in a session.

**Review by jacoblee93 — APPROVED** [review]
Thanks! Looks good

**Pull request merged**